Libplanet.Tests/Action/ActionEvaluatorTest.cs

```diff
@@ -0,0 +1,997 @@
+using Libplanet.Blockchain.Policies;
+using Libplanet.Blocks;
+using Libplanet.Crypto;
+using Libplanet.Store;
+using Libplanet.Tests.Common.Action;
+using Libplanet.Tests.Fixtures;
+using Libplanet.Tests.Store;
+using Libplanet.Tests.Store.Trie;
+using Libplanet.Tests.Tx;
+using Libplanet.Tx;
+using Serilog;
+using Xunit;
+using Xunit.Abstractions;
+using static Libplanet.Tests.TestUtils;
+
+namespace Libplanet.Tests.Action
+{
+    public class ActionEvaluatorTest
+    {
+        private static readonly AccountStateGetter _nullAccountStateGetter = address => null;
+        private static readonly AccountBalanceGetter _nullAccountBalanceGetter =
+            (address, currency) => new FungibleAssetValue(currency);
+
+        private readonly ILogger _logger;
+        private readonly BlockPolicy<DumbAction> _policy;
+        private readonly StoreFixture _storeFx;
+        private readonly TxFixture _txFx;
+
+        public ActionEvaluatorTest(ITestOutputHelper output)
+        {
+            Log.Logger = _logger = new LoggerConfiguration()
+                .MinimumLevel.Verbose()
+                .Enrich.WithThreadId()
+                .WriteTo.TestOutput(output)
+                .CreateLogger()
+                .ForContext<ActionEvaluatorTest>();
+
+            _policy = new BlockPolicy<DumbAction>(new MinerReward(1), maxBlockBytes: 50 * 1024);
+            _storeFx = new DefaultStoreFixture(memory: true, blockAction: _policy.BlockAction);
+            _txFx = new TxFixture(null);
+        }
+
+        [Fact]
+        public void Idempotent()
+        {
+            // NOTE: This test checks that blocks can be evaluated idempotently. Also it checks
+            // the action results in pre-evaluation step and in evaluation step are equal.
+            const int repeatCount = 2;
+            var signer = new PrivateKey();
+            var timestamp = DateTimeOffset.UtcNow;
+            var txAddress = signer.ToAddress();
+            var txs = new[]
+            {
+                Transaction<RandomAction>.Create(
+                    nonce: 0,
+                    privateKey: signer,
+                    genesisHash: null,
+                    actions: new[] { new RandomAction(txAddress), }),
+            };
+            var stateStore =
+                new TrieStateStore(new MemoryKeyValueStore(), new MemoryKeyValueStore());
+            var noStateRootBlock = TestUtils.MineGenesis(
+                timestamp: timestamp,
+                transactions: txs);
+            var stateRootBlock = TestUtils.MineGenesis(
+                timestamp: timestamp,
+                transactions: txs).AttachStateRootHash(stateStore, null);
+            StateGetter<RandomAction> nullStateGetter =
+                (address, hashDigest, stateCompleter) => null;
+            BalanceGetter<RandomAction> nullBalanceGetter =
+                (address, currency, hashDigest, fungibleAssetStateCompleter)
+                    => new FungibleAssetValue(currency);
```

Comment:
Would such null state/balance getters be frequently used?  If so, how about changing the visibility of `ActionEvaluator<T>._nullStateGetter` & `_nullBalanceGetter` into `public static readonly`?  (Or it might be better to define them in normal static methods instead.  If a method's signature is matched to a delegate type-wise, the compile treat the method as a value of the delegate type.)

Reply:
Went back and forth with this a little bit. Thought about exposing `ActionEvaluator<T>._nullStateGetter` and such as `public`, but then realized it would most likely only be for testing (and possibly for benchmarking). Now that I'm going over `ActionEvaluator.cs` again, `ActionEvaluator<T>._nullStateGetter` and `ActionEvaluator<T>._nullBalanceGetter` don't even get used. 🙃

I'll just remove them from `ActionEvaluator<T>`.

Reply:
If these are not useful to Libplanet users, but still useful for test/benchmark code how about turning them `internal static readonly`?

Reply:
Hmm... I'm not fully on board with the idea of making things `internal` for purely testing and/or benchmarking. However, from a slightly different point of view, I can get behind the idea of providing additional `internal` tools for developers using the library, including us, of course. I'll make the changes. 🙃

Reply:
On a second thought, I still think the answer is no. I'll make the changes regardless for now, since this can be separated to a different issue.

Reply:
Done. See 04ae033430c70f543115322e3f613da1a3d8a113.